sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -456,9 +474,52 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case DateType =>
       buildCast[Int](_, d => null)
     case TimestampType =>
-      buildCast[Long](_, t => timestampToLong(t).toInt)
-    case x: NumericType =>
-      b => x.numeric.asInstanceOf[Numeric[Any]].toInt(b)
+      buildCast[Long](_, t => {
+        val longValue = timestampToLong(t)
+        if (longValue == longValue.toInt) {
+          longValue.toInt
+        } else {
+          null
+        }
+      })
+    case ByteType =>
+      b => b.asInstanceOf[Byte].toInt
+    case ShortType =>
+      b => b.asInstanceOf[Short].toInt
+    case LongType =>
+      buildCast[Long](_, l =>
+        if (l == l.toInt) {
+          l.toInt
+        } else {
+          null
+        }
+      )
+    case x: FloatType =>
+      buildCast[Float](_, f =>
+        if (f <= Int.MaxValue && f >= Int.MinValue) {
+          f.toInt
```

Comment:
This question might be out-of-scope in this pr though... we don't need round-up here?
```
postgres=# select CAST(3.9 AS INT);
 int4 
------
    4
(1 row)
```

Reply:
Actually it is tricty to compare float and int if the float value is around Int.MaxValue or Int.MinValue.
```
scala> BigDecimal((Int.MaxValue + 1L).toString).toFloat <= Int.MaxValue
res1: Boolean = true
```
This is because `float` is also 32 bits long, and it uses 8 bit to represent the exponent field. While `int` is 32 bits long. So the comparison won't be accurate.

Reply:
Ur, I see. That behaivour looks interesting...
But, what I'm just worried about is that the query below has different output;
```
// this pr
scala> sql("SELECT CAST(float(3.5) AS INT)").show
+-------------------------------+
|CAST(CAST(3.5 AS FLOAT) AS INT)|
+-------------------------------+
|                              3|
+-------------------------------+

// postgresql
postgres=# SELECT CAST(float4 '3.5' AS INT);
 int4 
------
    4
(1 row)

// mysql (no float literal in mysql?)
mysql> SELECT CAST(3.5 AS SIGNED INT);
+-------------------------+
| CAST(3.5 AS SIGNED INT) |
+-------------------------+
|                       4 |
+-------------------------+
1 row in set (0.00 sec)
```
You mean that, since that comparison is inaccurate, this output difference is ok?

Reply:
Oh I thought you were talking about the corner case.
As per the SQL standard in section 9.2
> it is implementation- defined whether the approximation is obtained by rounding or by truncation
Spark always uses truncation. So I think we can simply follow the previous behavior on this. It is not the scope of this PR.

Reply:
yea, ok. thx for your check!